core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -406,11 +407,28 @@ public static Expression convert(Expression operand, Type fromType,
     if (toPrimitive != null) {
       if (fromPrimitive != null) {
         // E.g. from "float" to "double"
-        return Expressions.convert_(
+        if (toPrimitive == Primitive.BOOLEAN) {
+          return Expressions.convert_(
+              operand, toPrimitive.getPrimitiveClass());
+        }
+        return Expressions.convertChecked(
             operand, toPrimitive.getPrimitiveClass());
       }
-      if (fromNumber || fromBox == Primitive.CHAR) {
-        // Generate "x.shortValue()".
+      if (fromType == BigDecimal.class
+          && (toPrimitive == Primitive.INT
```

Comment:
could add a method to the `Primitive` class similar to `isNumeric()` in case another check like this is needed in the future, also reduces `if` complexity.